### ISSUE TITLE
[func-style] error when anonymous function is exported as default

### ISSUE THREAD

**Issue by jrencz** [eslint/eslint#5183]
Consider the following one-liner:

```js
export default function () {}
```

Eslint config (from demo):
ECMA features: 
- [x] modules

Rules:
- [x] func-style

Error is reported:

```
1:16 - Expected a function expression. (func-style)
```

Although the example given is obviously not a function expression there might be a way to define exception in the rule for the case when one want to have all functions created using function expressions but allow anonymous (or maybe even named?) function to be exported as default from a module.

It makes sense to do so for short modules that export one single function as default.

Without a way to express this exception one valid way to default-export a function is this:

```js
const init = function () {};

export {init as default};
```

which makes us define provisional variable which would be unneeded otherwise.

**Comment by eslintbot** [comment]
@jrencz Thanks for the issue! If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

**Comment by michaelficarra** [comment]
```js
export default (function () {});
```

There, now it's a function expression.

**Comment by jrencz** [comment]
@michaelficarra that's right, but then we're modifying the source just to meet the rule with its current possibilities.

Main benefit from using `func-style` is (as described in rule readme) disallowing functions to be hoisted. There's no other benefit from forcefully making this anonymous function a function expression than to make `func-style` pass - it won't be hoisted anyway in this very case.

**Comment by michaelficarra** [comment]
Well *technically* the binding *is* hoisted, but its name cannot be referenced from JavaScript code, so it's not observable. Anyway, if your goal is to never use function declarations, wrapping it in parentheses would indeed be the proper solution. If, however, your goal is to avoid a potential pitfall involving hoisting (I'm not sure what that pitfall is) then you are right, this one particular form should be an exception.

**Comment by nzakas** [comment]
This rule shouldn't apply for a default export.

**Comment by michaelficarra** [comment]
@nzakas Of course it should:

```js
f();
// ...
export default function f(){}
```

This is exactly the kind of thing this rule is meant to avoid. What we were discussing is whether it should apply to an unnamed default export.

**Comment by nzakas** [comment]
The rule isn't about enforcing a name or not, it's about the stylistic preference between these:

```js
function foo() {}
var foo = function(){};
```

Forcing people to put parentheses around or name a default exported function doesn't seem remotely related to that.

**Comment by michaelficarra** [comment]
@nzakas From the docs:

> The primary difference between function declarations and function expressions is that declarations are hoisted to the top of the scope in which they are defined, which allows you to write code that uses the function before the declaration. For example:
> 
> ```js
> doSomething();
> 
> function doSomething() {
>     // ...
> }
> ```
> 
> Although this code might seem like an error, it actually works fine because JavaScript engines hoist the function declarations to the top of the scope. That means this code is treated as if the declaration came before the invocation.
> 
> For function expressions, you must define the function before it is used, otherwise it causes an error.

The style is that they don't want to see a reference to a function above the definition of that function. So this rule should certainly apply for a default export of a named function declaration. The only remaining question is whether this rule should treat a default export of an unnamed function declaration the same way.

**Comment by nzakas** [comment]
> The style is that they don't want to see a reference to a function above the definition of that function

That's an inference that I don't think is 100% valid. We have another rule `no-use-before-define` that is a better indicator of this intent. This rule is just about style, and the docs explain the consequences of changing that style not as a rationalization for the rule but for the edification of the reader.

And yes, the question of this issue is about unnamed default export functions. My feeling is that these should be omitted from this rule.

**Comment by BYK** [comment]
@nzakas - Is the action item "This rule shouldn't apply for a default export.", thus make the rule skip on default exports?

**Comment by nzakas** [comment]
Yes

**Comment by michaelficarra** [comment]
:unamused: unnamed default exports

**Comment by nzakas** [comment]
Here's my concern: I just don't believe that people who want their functions to look like this:

```js
var foo = function(){};
```

are also going to want default exports to look like this:

```js
export default (function(){})
```

Nor do I think that forcing them to remove the ID makes any sense if they prefer function expressions because, again, this isn't about hoisting or about names, it's about the preferred style for writing functions as either an assignment or a declaration. In the case of default exports, you don't have a choice - it's always a declaration so long as its not surrounded by parentheses. Whether or not it has ID doesn't really matter.

I mistyped before because I forgot about the case of people preferring declaration vs. expression, I think we just omit all default exports as I don't think it makes sense for this rule to apply.

**Comment by michaelficarra** [comment]
@nzakas Then why not have an exception for all export forms? How is it different than preferring

```js
var f = function(){};
export f;
```

to

```js
export function f(){}
```

which this rule would currently enforce?

**Comment by nzakas** [comment]
@michaelficarra if I'm not mistaken, you can do this:

```js
export var f = function(){};
```

But you cannot do this:

```js
export default var f = function(){};
```

So it seems like checking nondefault exports is still a valid use case.

**Comment by michaelficarra** [comment]
Oh okay, so you just don't think users who prefer function expressions would want to use the parentheses in export default position. I can accept that.

**Comment by jrencz** [comment]
maybe this rule should just ignore exports and we need another rule just to handle exports?

**Comment by ababba15** [comment]
Would be cool to have some rule that allows to export like this:

```
export default (props) => (
    <View>
        <TextInput {...props} />
    </View>
)
```
